drivers/timer/cdns/timer.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2024, UNSW
+ * Copyright 2025, UNSW
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdint.h>
+#include <microkit.h>
+#include <sddf/timer/protocol.h>
+#include <sddf/util/util.h>
+#include <sddf/util/printf.h>
+#include <sddf/resources/device.h>
+
+__attribute__((__section__(".device_resources"))) device_resources_t device_resources;
+
+/*
+ * Zynq UltraScale+ MPSoC contains a timer (Cadence Triple Timer Clock) with 3 32-bit counters.
+ * Only 2 of the 3 timers are used in this driver. All 3 timers by default are connected to
+ * the LSBUS clock in the Low Power Domain, (controlled by LPD_LSBUS_CTRL).
+ * By default, this clock is using IOPLL clock as source, and divides down by 15 to 100MHz.
+ * See: https://docs.amd.com/r/en-US/ug1087-zynq-ultrascale-registers/LPD_LSBUS_CTRL-CRL_APB-Register.
+ * You can read the IOPLL clk value in u-boot using `clk dump` and view the source and divider value for
+ * LSBUS by reading LPD_LSBUS_CTRL register.
+ */
+/* TODO:
+ * More graceful error when user requests timeout greater than served by ZCU102_TIMER_MAX_TICKS (at 25MHz (downscaled), 32-bit counter thats ~2min 51.799s)
+ * Allow timeouts greater than UINT32_MAX ticks
+ */
+
+/* prescale: rate = clk_src/[2^(N+1)] */
+#define ZCU102_TIMER_PRESCALE_VALUE 0x1
+/* 100MHz on the default source clock, scaled down to 25MHz */
+#define ZCU102_TIMER_TICKS_PER_SECOND 0x17D7840
```

Comment:
Note to self: the clock frequency value should be wrapped in an ifdef, as this is board specific, set in u-boot. SImilar to the serial driver: https://github.com/au-ts/sddf/pull/413#discussion_r2071091046

Can't change this, until we have a clock driver. (ATM this is the 100MHz LSBUS clock freq, scaled down by a factor of 4 to 25MHz)